agera/src/main/java/com/google/android/agera/FunctionCompiler.java

```diff
@@ -32,15 +32,48 @@
 
 @SuppressWarnings({"unchecked, rawtypes"})
 final class FunctionCompiler implements FList, FItem {
+  private static final ThreadLocal<FunctionCompiler> compilers = new ThreadLocal<>();
+
   @NonNull
-  private final List<Function> functions;
+  static FunctionCompiler functionCompiler() {
+    FunctionCompiler compiler = compilers.get();
+    if (compiler == null) {
+      compiler = new FunctionCompiler();
+    } else {
+      // Remove compiler from the ThreadLocal to prevent reuse in the middle of a compilation.
+      // recycle(), called by compile(), will return the compiler here. ThreadLocal.set(null) keeps
+      // the entry (with a null value) whereas remove() removes the entry; because we expect the
+      // return of the compiler, don't use the heavier remove().
+      compilers.set(null);
+    }
+    return compiler;
+  }
+
+  @NonNull
+  private static Object[] recycle(@NonNull final FunctionCompiler compiler) {
+    final List<Object> functions = compiler.functions;
+    final Object[] chainedFunctions = functions.toArray(new Object[functions.size()]);
+    functions.clear();
+    compilers.set(compiler);
+    return chainedFunctions;
+  }
+
+  private static final int FUNCTION = 0;
```

Comment:
chance of calling this APPLY instead? this makes all 5 constants verbs and IMHO more readable.